feat: multi-display support (macOS)

vasusharma7 opened the pull request:
This PR contributes  to https://github.com/helmerapp/micro/issues/59 with following changes: 

* Added a submenu to system tray menu to choose display targets ; disabled for non-Mac platform !
 * By default chooses the first target returned by scap and sets that (usually primary display)
 * Hard-coded (4) number of displays supported as of now.
	* Why this is so ? : Window labels need to be unique and pre-declared constants in tauri capabilities.json file (to the best of my knowledge) and cannot be re-assigned. So, these are hardcoded to 4 till a better solution is found. :(

<img width="570" alt="image" src="https://github.com/helmerapp/micro/assets/40715071/bc467afd-7365-4b89-bec0-5d268d60e20c">

vasusharma7 commented:
@Pranav2612000 @clearlysid , PTAL

+ Some of the testing (+code refactoring) is underway. Still raised this PR to get early feedback ! 
+ Being new to rust, might have written some idiomatic code and first things which worked for me (:p), eager to learn any good practices which should be followed.
+ Please have a thorough look incase I missed something or disturbed existing functionality.

clearlysid commented:
Thanks for the PR!

A more efficient way to get multi-display to work is to just spawn the record button (and cropper, etc.) on the display that the mouse cursor is present on instead of making 4 croppers and 4 record buttons.

1. on hotkey press or tray icon clicked
2. check mouse cursor position
3. detect monitor cursor is on
4. spawn cropper (and record button) on the detected monitor instead of primary monitor

You can use the `monitor_from_point` method to implement this, docs here: https://docs.rs/tauri/2.0.0-beta.22/tauri/struct.App.html#method.monitor_from_point

vasusharma7 commented:
Yes that was my intention to begin with and I resonate with the process and all the steps you mentioned !

But there were a few issues ,firstly, e-spawning of windows is not possible ( that same label cant be reused even after destroying windows)

Then I  tried to use the same window spawned before  and just to reposition it on the correct monitor and change its inner width and other properties but certain properties were not modifiable after the window is built (.build() method is called ). 

I don't correctly remember the issue( I think it was the width not getting modified ) but spent some time on this and then I tried to ship an MVP with a static number of screens.

That being said , I know it's a bad code design and bad UX to have a fixed number of screens

I feel it's the time that I take a step back now and try the above things again.

Please give me some time and I will post an update soon.

Thanks

clearlysid commented:
>  certain properties were not modifiable after the window is built (.build() method is called ).

you can use the following methods:
1. set_size: https://docs.rs/tauri/2.0.0-beta.22/tauri/window/struct.Window.html#method.set_size
2. set_position: https://docs.rs/tauri/2.0.0-beta.22/tauri/window/struct.Window.html#method.set_position

vasusharma7 commented:
I think I saw this 
Let me zoom out and give this a try again .... 
Will post an update soon

vasusharma7 commented:
I have updated a few things ( few optimisations still pending ) ; however I don't have second monitor with me for 2 weeks ( in my hometown ) - so could not test it thoroughly - Still, I will try to do some hack to test it and post an update soon.

clearlysid commented on src-tauri/src/cropper/mod.rs:
```diff
@@ -40,40 +42,119 @@ fn set_transparency_and_level(win: tauri::WebviewWindow, level: u32) {
                 #[cfg(target_os = "macos")]
-                state: Some(WindowEffectState::Active),
+                {
+                    let position: LogicalPosition<f64> = unsafe {
```
Tauri should have built-in constructs for getting the monitor bounds. Let's avoid using the `core-graphics` crate for this otherwise we will have to write separate implementations for Windows and Linux as well.

vasusharma7 commented:
To be honest, I don't recall why I added this dependency on core graphics at first place. Me and Pranav had some initial conversations quite some time back and we realised taking this dependency from core graphics and now I don't recall why we came to this conclusion ( that was the reason behind my other initial PR on scap to expose CG deduced properties )

Anyways, I changed it to just take everything from tauri::Monitor including the position, size, scale factor etc.. now But again I couldn't test it to work on multiple monitors as I don't currently have access to those :(

clearlysid commented:
Closing this in favour of #96 which I've updated to support both Windows and Mac.
